### PR TITLE
[FIX] auth_ldap: LDAP Authentication error handling

### PR DESCRIPTION
**Pull request by manali-malpani**
**Description of the issue/feature this PR addresses:**

There a critical error in the authentication module of the Odoo application which makes it possible to bypass authentication and impersonate any user (including administrators) in the system.
As part of the ITSS 4.3.1 d., CWE-303 this is a critical error and needs to be resolved as these information leaks might be relevant from security perspective.

**Current behavior before PR:** 

Currently it is possible to bypass LDAP authentication due to logical error and improper handling of unexpected error conditions.

- As an example, the application is logging in users which are locked in the LDAP repository with any arbitrary password. This is happening because in this case the LDAP system is returning another error code than 49 (Invalid Credentials) and the application doesn’t handle the error properly. 
As a result an anonymous attacker can impersonate any user in the application by locking the account with numerous failed login attempts and then log in to the application with an arbitrary password.
Other LDAP error conditions such as LDAP server becomes unavailable during the middle of the authentication process would provide the same results. 

- Also if LDAP fails to authenticate for any reason other than INVALID_CREDENTIALS, it actually lets you in as though authentication was successful. For example, if the LDAP server isn't running or has a slow connection, the reason will be SERVER_DOWN or TIMEOUT. There are many other possibilities, but those two are obvious ones the code should anticipate.

**Desired behavior after PR is merged:**

The error handling of the authentication function once fixed:

- users cannot login to the system with an arbitrary password if their account is locked.
- There is proper error handling for LDAP TIMEOUT issues or if the LDAP server is down.
- An additional exception is added to catch any other exceptions.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by azmeuk — APPROVED** [review]

File: addons/auth_ldap/models/res_company_ldap.py [hunk 1]
```diff
@@ -118,10 +118,21 @@ def _authenticate(self, conf, login, password):
...
             conn.unbind()
-        except ldap.INVALID_CREDENTIALS:
+        except ldap.INVALID_CREDENTIALS as e:
+            _logger.error('User Id or password is incorrect: %s', e)
+            return False
+        except ldap.TIMEOUT as e:
+            _logger.error('LDAP TIMEOUT occurred: %s', e)
+            return False
+        except ldap.SERVER_DOWN as e:
+            _logger.error('LDAP server down exception occurred: %s', e)
```
Comment (bodedra) [inline addons/auth_ldap/models/res_company_ldap.py:128]:
```suggestion
            _logger.error('An LDAP server down exception occurred: %s', e)
```

File: addons/auth_ldap/models/res_company_ldap.py [hunk 2]
```diff
@@ -118,10 +118,21 @@ def _authenticate(self, conf, login, password):
...
             return False
         except ldap.LDAPError as e:
-            _logger.error('An LDAP exception occurred: %s', e)
+            _logger.error('LDAP error occurred: %s', e)
```
Comment (bodedra) [inline addons/auth_ldap/models/res_company_ldap.py:131]:
```suggestion
            _logger.error('An LDAP exception occurred: %s', e)
```

File: addons/auth_ldap/models/res_company_ldap.py [hunk 3]
```diff
@@ -118,10 +118,21 @@ def _authenticate(self, conf, login, password):
...
+            return False
+        except:
+            _logger.error('LDAP catch all exception occurred')
+            return False
```
Comment (bodedra) [inline addons/auth_ldap/models/res_company_ldap.py:135]: This is not a helpful log message.
```suggestion
```

**Review by bodedra — COMMENTED** [review]
@manali-malpani Minor suggestion.

File: addons/auth_ldap/models/res_company_ldap.py [hunk 4]
```diff
@@ -118,10 +118,21 @@ def _authenticate(self, conf, login, password):
             conn = self._connect(conf)
             conn.simple_bind_s(dn, to_text(password))
             conn.unbind()
-        except ldap.INVALID_CREDENTIALS:
+        except ldap.INVALID_CREDENTIALS as e:
+            _logger.error('User Id or password is incorrect: %s', e)
```
Comment (bodedra) [inline addons/auth_ldap/models/res_company_ldap.py:122]:
```suggestion
            _logger.error('An LDAP User Id or password is incorrect: %s', e)
```

**Review by manali-malpani — COMMENTED** [review]
Hello Bhavesh,

I have accepted the suggestions. That's correct! Sentences should have the right article usage to make it more meaningful. 

Thanks!

**Comment by bodedra** [comment]
Thanks Manali for changes. One more last change, would you please squash your commits? At the end, it should have two commits, 1) FIX and 2) CLA

**Review by bodedra — APPROVED** [review]
Thanks @manali-malpani for notice and fix it. LGTM :+1: 

@odony @Xavier-Do @xmo-odoo @Yajo This is related with LDAP security. So your valuable input would be much appreciate.

**Comment by odony** [comment]
Thanks for the PR. In the future, please follow the responsible disclosure procedure, as discussed on e.g. on https://github.com/odoo/odoo/security/policy and https://www.odoo.com/security-report so we can evaluate the severity of the issue before disclosing it.
Most projects have such a policy, and locating it should be the first thing you do when you're facing a security concern.

With regards to the PR, it should target the lowest supported version where the code is present, so in this case that means version 12.0. Patches are automatically foward-ported in later series, so there is no need for more than a single PR. You could reapply the patch on top of v12, force-push, and re-target the PR, or simply make a new one.

Regarding the error handling, there is no need for special-casing each exception, it's brittle (as we've seen), and requires writing useless log messages. Essentially the only thing you need is `return False` inside the `except ldap.LDAPError` clause.

File: addons/auth_ldap/models/res_company_ldap.py [hunk 5]
```diff
@@ -118,10 +118,18 @@ def _authenticate(self, conf, login, password):
...
             conn.unbind()
-        except ldap.INVALID_CREDENTIALS:
+        except ldap.INVALID_CREDENTIALS as e:
+            _logger.error('An LDAP User Id or password is incorrect: %s', e)
+            return False
+        except ldap.TIMEOUT as e:
+            _logger.error('An LDAP timeout exception occurred: %s', e)
+            return False
+        except ldap.SERVER_DOWN as e:
+            _logger.error('An LDAP server down exception occurred: %s', e)
```
Comment (odony) [inline addons/auth_ldap/models/res_company_ldap.py:128]: none of those changes are necessary. There are many more kinds of exception and we don't need a specific `except` clause for each. Just let them fall through to the `ldap.LDAPError` case, the log message will contain the exact error.

**Comment by bodedra** [comment]
@odony Thanks for your quick review, response and provide guidance on the reporting security concern. We will maintain privacy. Thanks again.

@manali-malpani Let's follow security policy and process further. And do we have any server log or any kind evidence of vulnerable? Let's connect separately.

**Review by yajo — APPROVED** [review]
I cannot say anything else that @odony didn't say already. 👍🏼

**Comment by manali-malpani** [comment]
Thanks @odony @bodedra for your comments. 

I will remove the  exception handling for timeout and server down. 
        except ldap.TIMEOUT as e:
            _logger.error('An LDAP timeout exception occurred: %s', e)
            return False
        except ldap.SERVER_DOWN as e:
            _logger.error('An LDAP server down exception occurred: %s', e)
            return False

And keep the logs if the user entered invalid credentials. And return the value False for LDAP error authentication and modify the PR.

**Comment by manali-malpani** [comment]
Hello @odony @Yajo 

I will close this pull request and drop the details in mail as discussed with @bodedra to security@odoo.com
However, even though this patch will be fixing the issue however as per the IBM Security guidelines the best practice is to positively confirm that the authentication was successful. Here the opposite is happening: the client is authenticating with the server and instead of testing for success it is trying to handle all possible errors. 

This would provide correct operation, however according to the principles of secure coding authentication functions must test for success explicitly instead of testing for pre-defined error conditions. For example:

# pseudo code
def auth(user, pass)
	successful_auth=False
	response=auth_with_external_user_repo(user,pass)
	if(response==success):
		successful_auth=True
	return successful_auth



I did little analysis on the LDAP module in Python.
https://www.python-ldap.org/en/python-ldap-3.3.0/reference/ldap.html#sending-ldap-requests

As per my analysis the documentation does not return with with any useful or success value.
However, I have been told the documentation is not accurate and it is possible to test for the success based on the return value.
Below is the source code for the LDAP module.

https://github.com/python-ldap/python-ldap/blob/master/Lib/ldap/ldapobject.py

def simple_bind_s(self,who=None,cred=None,serverctrls=None,clientctrls=None):
    """
    simple_bind_s([who='' [,cred='']]) -> 4-tuple
    """
    msgid = self.simple_bind(who,cred,serverctrls,clientctrls)
    resp_type, resp_data, resp_msgid, resp_ctrls = self.result3(msgid,all=1,timeout=self.timeout)
    return resp_type, resp_data, resp_msgid, resp_ctrls


Kindly pour your thoughts on the same and let me know if it is feasible to do so.

Thank You!

Kind regards,
Manali

**Comment by Yajo** [comment]
You have the patch for v12 in https://github.com/odoo/odoo/pull/72209. This one can be merged or can be closed.

**Comment by manali-malpani** [comment]
I will close this PR and create a new PR against ODOO version 12.0
Thank You for your inputs.

**Comment by odony** [comment]
> I will close this pull request and drop the details in mail (...)

At this point you could have just fixed the PR (e.g. by re-targeting to series 12.0 and squashing the commits), as @Yajo tried to do for you. 


> (...) Here the opposite is happening: the client is authenticating with the server and instead of testing for success it is trying to handle all possible errors.

I understand your concern, but that description isn't quite accurate. `python-ldap` is a library that abstracts the low-level procotol details because LDAP is complicated and most client stacks don't care about them. Yet behind the scene it does verify the result code returned by the server, and converts any error into an appropriate Python-level exception:
  https://github.com/python-ldap/python-ldap/blob/3.0/Modules/LDAPObject.c#L526-L527
(LDAPError is a function that will raise a Python exception)

Therefore, a Python client application using `python-ldap` is only supposed to handle the documented exceptions, and must consider "no exception' as a successful bind. This is not only the documented way of using the API, but it's also the only possible way, because the library does not return the actual LDAP result code (cfr the code linked above, and [`ldap_error(3)`](https://linux.die.net/man/3/ldap_error)). But by construction we know that the result was `LDAP_SUCCESS`, otherwise there would be an exception.

**Comment by odony** [comment]
As we can see, the special-case for INVALID_CREDENTIALS is not strictly necessary, its only merit currently is to handle the error logging differently, and actually to silence it. The authentication layer of Odoo will log successful/failed logins anyway, but because it supports multiple authentication methods, some of them may fail (and might report failures) even though another one succeeded. If each layer individually logs success/errors, we could have spurious messages in the logs.
In this way, the Odoo auth layers are similar to the Linux PAM system.

So I'd also remove the extra logging that was added, because it will create noise in the logs.

**Pull request closed without merging**